sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -1286,6 +1286,290 @@ __ESIMD_API void lsc_store2d(T *Ptr, unsigned SurfaceWidth,
+  lsc_load2d(config_2d_mem_access<T1> &payload);
+  template <typename T1, int N>
+  friend ESIMD_INLINE SYCL_ESIMD_FUNCTION void
+  lsc_store2d(config_2d_mem_access<T1> &payload, __ESIMD_NS::simd<T1, N> Data);
+
+};
+
+/// A variation of \c 2D stateless block load \c with parameters passed as
+/// \c config_2d_mem_access \c object
+/// Note: Compatibility with future hardware versions is not guaranteed.
+/// Note: No software mitigation for hardware bugs is possible for this function.
+/// \param T is the element data type
+/// \param N is the data size
+/// \param payload is \c config_2d_mem_access \c object holding all the data
+/// \return is a vector of type T and size N, where N is
+///  getNextPowerOf2(Height) * Width * NBlocks, if transposed
+///  getNextPowerOf2(Width) * Height * NBlocks, otherwise
+///
+template <typename T, int N>
+ESIMD_INLINE SYCL_ESIMD_FUNCTION __ESIMD_NS::simd<T, N>
+lsc_load2d(config_2d_mem_access<T> &payload) {
+  __ESIMD_NS::simd<T, N> oldDst;
+  constexpr uint32_t exDesc = 0x0;
+  constexpr uint32_t desc = 0x28A0403;
+  constexpr uint8_t execSize = 0x0;
+  constexpr uint8_t sfid = 0xF;
+  constexpr uint8_t numSrc0 = 0x1;
+  constexpr uint8_t numSrc1 = 0x0;
+  constexpr uint8_t numDst = (N * sizeof(T)) / 64;
+  return raw_send_load(oldDst, payload.get_raw_data(), exDesc, desc, execSize,
+                       sfid, numSrc0, numDst);
+}
+
+/// A variation of \c 2D stateless block store \c with parameters passed as
+/// \c config_2d_mem_access \c object
+/// Note: Compatibility with future hardware versions is not guaranteed.
+/// Note: No software mitigation for hardware bugs is possible for this function.
+/// \param T is the element data type
+/// \param N is the data size
+/// \param payload is \c config_2d_mem_access \c object holding all the data
+/// \param Data is the data to be stored.
+///
+template <typename T, int N>
+ESIMD_INLINE SYCL_ESIMD_FUNCTION void
+lsc_store2d(config_2d_mem_access<T> &payload, __ESIMD_NS::simd<T, N> Data) {
+
+  constexpr uint32_t exDesc = 0x0;
+  constexpr uint32_t desc = 0x2040407;
+  constexpr uint8_t execSize = 0x0;
+  constexpr uint8_t sfid = 0xF;
+  constexpr uint8_t numSrc0 = 0x1;
+  constexpr uint8_t numSrc1 = (N * sizeof(T)) / 64;
+  constexpr uint8_t numDst = 0x0;
+  raw_sends_store(payload.get_raw_data(), Data, exDesc, desc, execSize, sfid,
+                  numSrc0, numSrc1);
+}
+
+/// Computes the data size for 2d block load or store.
+/// \param NBlocks is the number of blocks
+/// \param Width is the block width in number of elements
+/// \param Height is the block height in number of elements
+/// \param Transposed is Transposed or not
+/// \return data size in blocks
+
+template <int NBlocks, int Height, int Width, bool Transposed>
+constexpr int get_block_2d_data_size() {
+  if constexpr (Transposed)
+    return __ESIMD_DNS::getNextPowerOf2<Height>() * Width * NBlocks;
+  return __ESIMD_DNS::getNextPowerOf2<Width>() * Height * NBlocks;
```

Comment:
This is not correct in common case. When `NBlocks` is greater, then 1, each block should be GRF-aligned.